### Анализ датасета №1 "Продажи игр"

### Выполнил студент: Глазков Тимур Михайлович

# Импорт библиотек

In [5]:
import numpy as np
import pandas as pd
import plotly.express as px

%matplotlib inline

# Загрузка и предобработка данных

In [6]:
games_data = pd.read_csv("vgsales.csv")
games_data.head()

Rank                      Name Platform    Year         Genre Publisher   
0     1                Wii Sports      Wii  2006.0        Sports  Nintendo  \
1     2         Super Mario Bros.      NES  1985.0      Platform  Nintendo   
2     3            Mario Kart Wii      Wii  2008.0        Racing  Nintendo   
3     4         Wii Sports Resort      Wii  2009.0        Sports  Nintendo   
4     5  Pokemon Red/Pokemon Blue       GB  1996.0  Role-Playing  Nintendo   

   NA_Sales  EU_Sales  JP_Sales  Other_Sales  Global_Sales  
0     41.49     29.02      3.77         8.46         82.74  
1     29.08      3.58      6.81         0.77         40.24  
2     15.85     12.88      3.79         3.31         35.82  
3     15.75     11.01      3.28         2.96         33.00  
4     11.27      8.89     10.22         1.00         31.37

Копирование датасета для безопасной работы

In [7]:
games_df = games_data.copy()
games_df.info()
games_df.sample(8)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16598 entries, 0 to 16597
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          16598 non-null  int64  
 1   Name          16598 non-null  object 
 2   Platform      16598 non-null  object 
 3   Year          16327 non-null  float64
 4   Genre         16598 non-null  object 
 5   Publisher     16540 non-null  object 
 6   NA_Sales      16598 non-null  float64
 7   EU_Sales      16598 non-null  float64
 8   JP_Sales      16598 non-null  float64
 9   Other_Sales   16598 non-null  float64
 10  Global_Sales  16598 non-null  float64
dtypes: float64(6), int64(1), object(4)
memory usage: 1.4+ MB


Rank                                    Name Platform    Year   
8637    8639  LEGO Legends of Chima: Laval's Journey      PSV  2013.0  \
9807    9809          Showdown: Legends of Wrestling       XB  2004.0   
3397    3399                          NFL Blitz 2001       PS  2000.0   
82        83                          FIFA Soccer 13      PS3  2012.0   
5112    5114              Rampage: Total Destruction      PS2  2006.0   
1708    1710           Pro Yakyuu Family Stadium '88      NES  1988.0   
1748    1750      LEGO Star Wars III: The Clone Wars       DS  2011.0   
13033  13035                              Top Spin 2       DS  2006.0   

           Genre                               Publisher  NA_Sales  EU_Sales   
8637   Adventure  Warner Bros. Interactive Entertainment      0.02      0.09  \
9807    Fighting                   Acclaim Entertainment      0.09      0.03   
3397      Sports                            Midway Games      0.33      0.22   
82        Action                         Electronic Arts      1.06      5.05   
5112      Action                            Midway Games      0.18      0.14   
1708      Sports                      Namco Bandai Games      0.10      0.00   
1748      Action                               LucasArts      0.66      0.39   
13033     Sports                    Take-Two Interactive      0.05      0.00   

       JP_Sales  Other_Sales  Global_Sales  
8637       0.00         0.04          0.16  
9807       0.00         0.00          0.12  
3397       0.00         0.04          0.59  
82         0.13         2.01          8.24  
5112       0.00         0.05          0.37  
1708       1.08         0.00          1.18  
1748       0.00         0.11          1.16  
13033      0.00         0.00          0.05

Преобразование колонок

In [8]:
games_df.columns = [*map(lambda x: x.lower(), games_df.columns)]
games_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16598 entries, 0 to 16597
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   rank          16598 non-null  int64  
 1   name          16598 non-null  object 
 2   platform      16598 non-null  object 
 3   year          16327 non-null  float64
 4   genre         16598 non-null  object 
 5   publisher     16540 non-null  object 
 6   na_sales      16598 non-null  float64
 7   eu_sales      16598 non-null  float64
 8   jp_sales      16598 non-null  float64
 9   other_sales   16598 non-null  float64
 10  global_sales  16598 non-null  float64
dtypes: float64(6), int64(1), object(4)
memory usage: 1.4+ MB


Удаление дубликатов

In [9]:
games_df.drop_duplicates()
games_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16598 entries, 0 to 16597
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   rank          16598 non-null  int64  
 1   name          16598 non-null  object 
 2   platform      16598 non-null  object 
 3   year          16327 non-null  float64
 4   genre         16598 non-null  object 
 5   publisher     16540 non-null  object 
 6   na_sales      16598 non-null  float64
 7   eu_sales      16598 non-null  float64
 8   jp_sales      16598 non-null  float64
 9   other_sales   16598 non-null  float64
 10  global_sales  16598 non-null  float64
dtypes: float64(6), int64(1), object(4)
memory usage: 1.4+ MB


Все значения столбца "rank", следовательно его можно сделать индексом

In [10]:
display(games_df['rank'].nunique(), games_df['rank'].count())
games_df.index = games_df['rank']
games_df = games_df.drop('rank', axis=1)
games_df.head()

16598

16598

name platform    year         genre publisher   
rank                                                                      
1                   Wii Sports      Wii  2006.0        Sports  Nintendo  \
2            Super Mario Bros.      NES  1985.0      Platform  Nintendo   
3               Mario Kart Wii      Wii  2008.0        Racing  Nintendo   
4            Wii Sports Resort      Wii  2009.0        Sports  Nintendo   
5     Pokemon Red/Pokemon Blue       GB  1996.0  Role-Playing  Nintendo   

      na_sales  eu_sales  jp_sales  other_sales  global_sales  
rank                                                           
1        41.49     29.02      3.77         8.46         82.74  
2        29.08      3.58      6.81         0.77         40.24  
3        15.85     12.88      3.79         3.31         35.82  
4        15.75     11.01      3.28         2.96         33.00  
5        11.27      8.89     10.22         1.00         31.37

Описательная статистика

In [11]:
display(games_df.describe())
games_df.describe(include='object')

year      na_sales      eu_sales      jp_sales   other_sales   
count  16327.000000  16598.000000  16598.000000  16598.000000  16598.000000  \
mean    2006.406443      0.264667      0.146652      0.077782      0.048063   
std        5.828981      0.816683      0.505351      0.309291      0.188588   
min     1980.000000      0.000000      0.000000      0.000000      0.000000   
25%     2003.000000      0.000000      0.000000      0.000000      0.000000   
50%     2007.000000      0.080000      0.020000      0.000000      0.010000   
75%     2010.000000      0.240000      0.110000      0.040000      0.040000   
max     2020.000000     41.490000     29.020000     10.220000     10.570000   

       global_sales  
count  16598.000000  
mean       0.537441  
std        1.555028  
min        0.010000  
25%        0.060000  
50%        0.170000  
75%        0.470000  
max       82.740000

name platform   genre        publisher
count                         16598    16598   16598            16540
unique                        11493       31      12              578
top     Need for Speed: Most Wanted       DS  Action  Electronic Arts
freq                             12     2163    3316             1351

Предобработка признака "year"

In [12]:
display(games_df[games_df['year'].isna()])
games_df['year'] = games_df['year'].fillna(games_df['year'].median())
games_df['year'] = games_df['year'].astype(np.int16)
games_df['year'].dtype

name platform  year         genre   
rank                                                              
180                Madden NFL 2004      PS2   NaN        Sports  \
378               FIFA Soccer 2004      PS2   NaN        Sports   
432     LEGO Batman: The Videogame      Wii   NaN        Action   
471     wwe Smackdown vs. Raw 2006      PS2   NaN      Fighting   
608                 Space Invaders     2600   NaN       Shooter   
...                            ...      ...   ...           ...   
16310                Freaky Flyers       GC   NaN        Racing   
16330                    Inversion       PC   NaN       Shooter   
16369  Hakuouki: Shinsengumi Kitan      PS3   NaN     Adventure   
16430                 Virtua Quest       GC   NaN  Role-Playing   
16496                   The Smurfs      3DS   NaN        Action   

                                    publisher  na_sales  eu_sales  jp_sales   
rank                                                                          
180                           Electronic Arts      4.26      0.26      0.01  \
378                           Electronic Arts      0.59      2.36      0.04   
432    Warner Bros. Interactive Entertainment      1.86      1.02      0.00   
471                                       NaN      1.57      1.02      0.00   
608                                     Atari      2.36      0.14      0.00   
...                                       ...       ...       ...       ...   
16310                                 Unknown      0.01      0.00      0.00   
16330                      Namco Bandai Games      0.01      0.00      0.00   
16369                                 Unknown      0.01      0.00      0.00   
16430                                 Unknown      0.01      0.00      0.00   
16496                                 Unknown      0.00      0.01      0.00   

       other_sales  global_sales  
rank                              
180           0.71          5.23  
378           0.51          3.49  
432           0.29          3.17  
471           0.41          3.00  
608           0.03          2.53  
...            ...           ...  
16310         0.00          0.01  
16330         0.00          0.01  
16369         0.00          0.01  
16430         0.00          0.01  
16496         0.00          0.01  

[271 rows x 10 columns]

dtype('int16')

Предобработка признаков продаж

In [13]:
sales_cols = games_df.columns[-5:]
print(sales_cols)
games_df[sales_cols] = games_df[sales_cols].apply(lambda col: col.astype(np.float16), axis=1)
games_df[sales_cols].describe()

Index(['na_sales', 'eu_sales', 'jp_sales', 'other_sales', 'global_sales'], dtype='object')


na_sales      eu_sales      jp_sales   other_sales  global_sales
count  16598.000000  16598.000000  16598.000000  16598.000000  16598.000000
mean       0.264648      0.146729      0.077820      0.048065      0.537598
std        0.816895      0.505371      0.309326      0.188599      1.555664
min        0.000000      0.000000      0.000000      0.000000      0.010002
25%        0.000000      0.000000      0.000000      0.000000      0.059998
50%        0.080017      0.020004      0.000000      0.010002      0.170044
75%        0.239990      0.109985      0.040009      0.040009      0.469971
max       41.500000     29.015625     10.218750     10.570312     82.750000

Так как курс доллара 1980 года менялся, под воздействием инфляции, имеет смысл сделать поправку на неё
1. Подгрузка и приобразование датасета инфляции доллара

In [14]:
infl_df = pd.read_csv('inflation_data.csv')
infl_df['amount'] = infl_df['amount'] / infl_df['amount'].iloc[-1]
infl_df.index = infl_df['year']
infl_df.drop('year', axis=1, inplace=True)
infl_df.tail(10)

amount  inflation rate
year                          
2014  0.769315            0.02
2015  0.770407            0.00
2016  0.779962            0.01
2017  0.796615            0.02
2018  0.816544            0.02
2019  0.831013            0.02
2020  0.841114            0.01
2021  0.880699            0.05
2022  0.951133            0.08
2023  1.000000            0.05

Терерь он отражает отношение современного курса доллара к курсу за каждый год

2. Функция коонвертации доллара по году

In [15]:
def convert_amount(sample: pd.Series) -> pd.Series:
    year = sample['year']
    multiplier = infl_df['amount'].loc[year]
    sample[sales_cols] *= multiplier
    return sample

Применение функции к каждой строке датасета (не запускать дважды)

In [16]:
games_df.apply(convert_amount, axis=1)

name platform  year   
rank                                                                     
1                                            Wii Sports      Wii  2006  \
2                                     Super Mario Bros.      NES  1985   
3                                        Mario Kart Wii      Wii  2008   
4                                     Wii Sports Resort      Wii  2009   
5                              Pokemon Red/Pokemon Blue       GB  1996   
...                                                 ...      ...   ...   
16596                Woody Woodpecker in Crazy Castle 5      GBA  2002   
16597                     Men in Black II: Alien Escape       GC  2003   
16598  SCORE International Baja 1000: The Official Game      PS2  2008   
16599                                        Know How 2       DS  2010   
16600                                  Spirits & Spells      GBA  2003   

              genre   publisher   na_sales   eu_sales  jp_sales  other_sales   
rank                                                                           
1            Sports    Nintendo  27.190827  19.011057  2.469799     5.543612  \
2          Platform    Nintendo  10.169008   1.252001  2.381056     0.269286   
3            Racing    Nintendo  11.091334   9.014100  2.651206     2.316389   
4            Sports    Nintendo  10.981572   7.675117  2.286466     2.064492   
5      Role-Playing    Nintendo   5.749053   4.533903  5.211200     0.509965   
...             ...         ...        ...        ...       ...          ...   
16596      Platform       Kemco   0.005849   0.000000  0.000000     0.000000   
16597       Shooter  Infogrames   0.005980   0.000000  0.000000     0.000000   
16598        Racing  Activision   0.000000   0.000000  0.000000     0.000000   
16599        Puzzle    7G//AMES   0.000000   0.007089  0.000000     0.000000   
16600      Platform     Wanadoo   0.005980   0.000000  0.000000     0.000000   

       global_sales  
rank                 
1         54.217854  
2         14.075962  
3         25.057995  
4         23.009009  
5         16.000137  
...             ...  
16596      0.005849  
16597      0.005980  
16598      0.006998  
16599      0.007089  
16600      0.005980  

[16598 rows x 10 columns]

Предобработка категориальных признаков

In [17]:
categorical_cols = ['platform', 'genre']
for col in categorical_cols:
    games_df[col] = games_df[col].astype('category')
    
games_df[categorical_cols].info()
games_df[categorical_cols].sample(8)

<class 'pandas.core.frame.DataFrame'>
Index: 16598 entries, 1 to 16600
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   platform  16598 non-null  category
 1   genre     16598 non-null  category
dtypes: category(2)
memory usage: 163.8 KB


platform      genre
rank                     
5143       PSP     Sports
15006      PS2  Adventure
5298        DS   Fighting
6           GB     Puzzle
4317       Wii  Adventure
16004       DS       Misc
3549       PS2   Strategy
6030       N64     Sports

Предобработка признака "publisher" в категориальный

In [18]:
MAX_CATEGORY_SIZE = 150

display(games_df['publisher'].value_counts().iloc[:MAX_CATEGORY_SIZE-1])
top_publishers = games_df['publisher'].value_counts().iloc[:MAX_CATEGORY_SIZE-1].index
replace_publisher = lambda publ: publ if publ in top_publishers else "Other"
games_df['publisher'] = games_df['publisher'].apply(replace_publisher)
games_df['publisher'] = games_df['publisher'].astype('category')
games_df['publisher'].value_counts()

publisher
Electronic Arts                 1351
Activision                       975
Namco Bandai Games               932
Ubisoft                          921
Konami Digital Entertainment     832
                                ... 
Brash Entertainment               10
Cave                              10
Xplosiv                           10
Ackkstudios                       10
Sunsoft                           10
Name: count, Length: 149, dtype: int64

publisher
Electronic Arts        1351
Other                  1237
Activision              975
Namco Bandai Games      932
Ubisoft                 921
                       ... 
Ackkstudios              10
Brash Entertainment      10
Cave                     10
Xplosiv                  10
Sunsoft                  10
Name: count, Length: 150, dtype: int64

# Сравнение исходного и обработанного датасетов

In [19]:
display(games_data.info())
display(games_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16598 entries, 0 to 16597
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Rank          16598 non-null  int64  
 1   Name          16598 non-null  object 
 2   Platform      16598 non-null  object 
 3   Year          16327 non-null  float64
 4   Genre         16598 non-null  object 
 5   Publisher     16540 non-null  object 
 6   NA_Sales      16598 non-null  float64
 7   EU_Sales      16598 non-null  float64
 8   JP_Sales      16598 non-null  float64
 9   Other_Sales   16598 non-null  float64
 10  Global_Sales  16598 non-null  float64
dtypes: float64(6), int64(1), object(4)
memory usage: 1.4+ MB


None

<class 'pandas.core.frame.DataFrame'>
Index: 16598 entries, 1 to 16600
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   name          16598 non-null  object  
 1   platform      16598 non-null  category
 2   year          16598 non-null  int16   
 3   genre         16598 non-null  category
 4   publisher     16598 non-null  category
 5   na_sales      16598 non-null  float16 
 6   eu_sales      16598 non-null  float16 
 7   jp_sales      16598 non-null  float16 
 8   other_sales   16598 non-null  float16 
 9   global_sales  16598 non-null  float16 
dtypes: category(3), float16(5), int16(1), object(1)
memory usage: 525.6+ KB


None

Как результат:
1. Восстановлены все пропущенные значения
2. Признаки преведены к корректным типам
3. Показатели продаж скорректированы относительно инфляции доллара
3. Количество занимаемой памяти сократилось приблизительно в 3 раза (1.4 MB -> 525.6 KB)

# Итоговая описательная статистика по численным и категориальным признакам

In [20]:
display(games_df.describe().round(4))
games_df.describe(include=['category', 'object'])

year    na_sales    eu_sales    jp_sales  other_sales   
count  16598.0000  16598.0000  16598.0000  16598.0000   16598.0000  \
mean    2006.4161      0.2646      0.1467      0.0778       0.0481   
std        5.7817      0.8169      0.5054      0.3093       0.1886   
min     1980.0000      0.0000      0.0000      0.0000       0.0000   
25%     2003.0000      0.0000      0.0000      0.0000       0.0000   
50%     2007.0000      0.0800      0.0200      0.0000       0.0100   
75%     2010.0000      0.2400      0.1100      0.0400       0.0400   
max     2020.0000     41.5000     29.0156     10.2188      10.5703   

       global_sales  
count    16598.0000  
mean         0.5376  
std          1.5557  
min          0.0100  
25%          0.0600  
50%          0.1700  
75%          0.4700  
max         82.7500

name platform   genre        publisher
count                         16598    16598   16598            16598
unique                        11493       31      12              150
top     Need for Speed: Most Wanted       DS  Action  Electronic Arts
freq                             12     2163    3316             1351

Отсюда видно, что:
* Большинство видео игр было выпущено в период с 2000 по 2010 года
* 50% рынка сконцентрировано в Северной Америке
* Продажи большинства видео игр не превышают 500 000 $, но при этом есть выбросы в виде игр с феноменальным успехо продаж, превыщающие среднее и медианное значение в десяти раз
* Жанр "Action" является самым популярным, так как встречается примерно в 20% выборки, а наиболее популярная платформа "DS" - примерно в 12%

# Визуализация

In [21]:
bar_data = games_df.groupby('genre', as_index=False)['global_sales'].count()\
                            .sort_values(['global_sales'], ascending=False)

fig = px.bar(
    data_frame=bar_data,
    x='genre',
    y='global_sales',
    color='genre',
    text='global_sales',
    orientation='v',
    height=500,
    width=1000,
    title='Топ жанров по количетву выпущенных игр',
    labels=dict(genre="Жанр", global_sales="Количество игр")
)


fig.update_xaxes(tickangle=-25)
fig.show()

In [22]:
bar_data = games_df.groupby('genre', as_index=False)['global_sales'].sum()\
                            .sort_values(['global_sales'], ascending=False)
bar_data['global_sales'] = bar_data['global_sales'].round()

fig = px.bar(
    data_frame=bar_data,
    x='genre',
    y='global_sales',
    color='genre',
    text='global_sales',
    orientation='v',
    height=500,
    width=1000,
    title='Топ жанров по сумме продаж',
    labels=dict(genre="Жанр", global_sales="Миллионов ($)")
)

fig.update_xaxes(tickangle=-25)
fig.show()

In [23]:
bar_data = games_df.groupby('year', as_index=False)['global_sales'].count()\
                            .sort_values(['global_sales'], ascending=False)
bar_data['global_sales'] = bar_data['global_sales'].round()

fig = px.bar(
    data_frame=bar_data,
    x='year',
    y='global_sales',
    color='year',
    text='global_sales',
    orientation='v',
    height=500,
    width=1000,
    title='Количетво выпущенных игр по годам',
    labels=dict(year="Год", global_sales="Количество игр"),
    range_x=(1979, 2017)
)

fig.show()

In [24]:
bar_data = games_df.groupby('year', as_index=False)['global_sales'].sum()\
                            .sort_values(['global_sales'], ascending=False)
bar_data['global_sales'] = bar_data['global_sales'].round()

fig = px.bar(
    data_frame=bar_data,
    x='year',
    y='global_sales',
    color='year',
    text='global_sales',
    orientation='v',
    height=500,
    width=1000,
    title='Сумма продаж по годам',
    labels=dict(year="Год", global_sales="Миллионы ($)"),
    range_x=(1979, 2017)
)

fig.show()

In [25]:
bar_data = games_df.groupby(['year', 'genre'], as_index=False)['global_sales'].count()\
                            .sort_values(['global_sales'], ascending=False)
top_years = tuple(games_df.groupby(['year'])['global_sales'].count().nlargest(5).index)
bar_data['global_sales'] = bar_data['global_sales'].round()
bar_data = bar_data[bar_data['year'].isin(top_years)]

fig = px.bar(
    data_frame=bar_data,
    x='year',
    y='global_sales',
    barmode='group',
    color='genre',
    text='global_sales',
    orientation='v',
    height=500,
    width=1000,
    title='Распределение количества игр по жанрам на пике релизов видео игр (ТОП-5 лет)',
    labels=dict(year="Год", global_sales="Количество игр", genre='Жанр'),
)
fig.show()

In [26]:
bar_data = games_df.groupby(['year', 'genre'], as_index=False)['global_sales'].sum()\
                            .sort_values(['global_sales'], ascending=False)
top_years = tuple(games_df.groupby(['year'])['global_sales'].sum().nlargest(5).index)
bar_data['global_sales'] = bar_data['global_sales'].round()
bar_data = bar_data[bar_data['year'].isin(top_years)]

fig = px.bar(
    data_frame=bar_data,
    x='year',
    y='global_sales',
    barmode='group',
    color='genre',
    text='global_sales',
    orientation='v',
    height=500,
    width=1000,
    title='Распределение суммы продаж по жанрам на пике продаж видео игр (ТОП-5 лет)',
    labels=dict(year="Год", global_sales="Миллионы ($)", genre='Жанр'),
)
fig.show()

In [27]:
bar_data = games_df.groupby(['platform'], as_index=False)['global_sales'].count()\
                            .nlargest(10, 'global_sales')

fig = px.bar(
    data_frame=bar_data,
    x='platform',
    y='global_sales',
    text='global_sales',
    orientation='v',
    height=500,
    width=1000,
    title='ТОП-10 платформ по суммарному количеству игр',
    labels=dict(platform="Платформа", global_sales="Количество игр"),
)
fig.show()

In [28]:
bar_data = games_df.groupby(['platform'], as_index=False)['global_sales'].sum()\
                            .nlargest(10, 'global_sales')
bar_data['global_sales'] = bar_data['global_sales'].round()


fig = px.bar(
    data_frame=bar_data,
    x='platform',
    y='global_sales',
    text='global_sales',
    orientation='v',
    height=500,
    width=1000,
    title='ТОП-10 платформ по продажам',
    labels=dict(platform="Платформа", global_sales="Миллионы ($)"),
)
fig.show()

In [29]:
bar_data = games_df.groupby(['publisher'], as_index=False)['global_sales'].count()\
                            .nlargest(11, 'global_sales')
bar_data = bar_data[bar_data['publisher'] != 'Other']

fig = px.bar(
    data_frame=bar_data,
    x='publisher',
    y='global_sales',
    text='global_sales',
    orientation='v',
    height=500,
    width=1000,
    title='ТОП-10 издателей по суммарному количеству игр',
    labels=dict(publisher="Издатель", global_sales="Количество игр"),
)

fig.update_xaxes(tickangle=-30)
fig.show()

In [30]:
bar_data = games_df.groupby(['publisher'], as_index=False)['global_sales'].sum()\
                            .nlargest(11, 'global_sales')
bar_data = bar_data[bar_data['publisher'] != 'Other']
bar_data['global_sales'] = bar_data['global_sales'].round()


fig = px.bar(
    data_frame=bar_data,
    x='publisher',
    y='global_sales',
    text='global_sales',
    orientation='v',
    height=500,
    width=1000,
    title='ТОП-10 издателей по продажам',
    labels=dict(publisher="Издатель", global_sales="Миллионы ($)"),
)

fig.update_xaxes(tickangle=-30)
fig.show()

In [31]:
bar_data = games_df.groupby(['name'], as_index=False)['global_sales'].sum()\
                            .nlargest(10, 'global_sales')
bar_data['global_sales'] = bar_data['global_sales'].round()


fig = px.bar(
    data_frame=bar_data,
    x='name',
    y='global_sales',
    text='global_sales',
    orientation='v',
    height=500,
    width=1000,
    title='ТОП-10 самых продаваемых игр за всю историю',
    labels=dict(name="Название", global_sales="Миллионы ($)"),
)

fig.update_xaxes(tickangle=-30)
fig.show()

In [32]:
pie_data = pd.DataFrame(games_df.sum(numeric_only=True)[sales_cols[:-1]]).T
pie_data.columns = ['Сев. Америка', 'Европа', 'Япония', 'Другое']
pie_data = pie_data.round().T
pie_data.reset_index(inplace=True)
pie_data.columns = ['name', 'sum']

fig = px.pie(
    pie_data,
    height=800,
    width=1000,
    values='sum',
    names='name',
    title='Распределение продаж по регионам',
    labels=dict(name="Регион", sum="Миллионы ($)")
)

fig.show()

In [34]:
heatmap_data = games_df.groupby('genre', as_index=False)\
                       .sum(numeric_only=True)[['genre']+[*sales_cols[:-1]]]
heatmap_data.index = heatmap_data['genre']
heatmap_data.drop('genre', axis=1, inplace=True)
heatmap_data.columns = ['Сев. Америка', 'Европа', 'Япония', 'Другое']
heatmap_data = heatmap_data.round().T

fig = px.imshow(
    heatmap_data,
    text_auto=True,
    height=500,
    width=1200,
    aspect="auto",
    title='Распределение продаж по жанрам и регионам',
    labels=dict(x="Жанр", color="Миллионы ($)", y="Регион")
)

fig.update_xaxes(tickangle=-30)
fig.show()

In [35]:
heatmap_data = games_df.groupby('platform', as_index=False)\
                       .sum(numeric_only=True)[['platform']+[*sales_cols[:-1]]]
heatmap_data.index = heatmap_data['platform']
heatmap_data.drop('platform', axis=1, inplace=True)
heatmap_data.columns = ['Сев. Америка', 'Европа', 'Япония', 'Другое']
heatmap_data = heatmap_data.round().T

fig = px.imshow(
    heatmap_data,
    text_auto=True,
    height=500,
    width=1200,
    aspect="auto",
    title='Распределение продаж по платформам и регионам',
    labels=dict(x="Платформа", color="Миллионы ($)", y="Регион")
)

fig.update_xaxes(tickangle=-30)
fig.show()

In [39]:
heatmap_data = games_df.groupby('year', as_index=False)\
                       .sum(numeric_only=True)[['year']+[*sales_cols[:-1]]]
heatmap_data.index = heatmap_data['year']
heatmap_data.drop('year', axis=1, inplace=True)
heatmap_data.columns = ['Сев. Америка', 'Европа', 'Япония', 'Другое']
heatmap_data = heatmap_data.round().T

fig = px.imshow(
    heatmap_data,
    text_auto=True,
    height=500,
    width=1400,
    aspect="auto",
    title='Распределение продаж по годам и регионам',
    labels=dict(x="Год", color="Миллионы ($)", y="Регион"),
)

fig.show()

In [ ]:
heatmap_data = games_df.corr(numeric_only=True).round(2)

fig = px.imshow(
    heatmap_data,
    text_auto=True,
    height=800,
    width=800,
    aspect="auto",
    title='Матрица корреляции признаков',
    labels=dict(color="Корреляция")
)

fig.update_xaxes(tickangle=-30)
fig.show()

Вывод по матрице корреляции признаков:
Так как датасет содержит преимущественно категориальные признаки, а числовые признаки (за исключением года) являются значениями продаж, сгруппированными по регионам, логично предположить, что между ними будет наблюдаться линейная зависимость, что подтверждается матрицей корреляции

# Гипотеза: 
#### Eсть жанр игр (возможно в связке с платформой или что-то еще), который продается лучше.

# Цель: 
#### Hайти наиболее перспективный жанр игры для потенциальной разработки.

Для подтверждения гипотезы и решения технической задачи поиска (цели), проведена группировка датасета по признакам ("genre" и "platform") в различных вариациях

В выборку включены наблюдения с 2007 для обеспечения большей актуальности трендов рынка

На древовидных графиках представленны распределения количества и сумарных продаж видео игр, сгруппированные по признакам в различных вариациях, для выявления популярности и продаваемости продукта в зависимости от признаков

In [ ]:
YEAR_FROM = 2007

treemap_count_data = games_df[games_df['year'] >= YEAR_FROM].groupby(['genre', 'platform'], 
                                                                as_index=False)['global_sales'].count()

fig = px.treemap(
    data_frame=treemap_count_data,
    path=['genre', 'platform'],
    values='global_sales',
    height=800,
    width=1000,
    title=f'Распределение количества игр по жанрам и платформам c {YEAR_FROM} года',
    labels=dict(parent="Жанр", global_sales="Количеество", labels='Платформа')
)
fig.show()

Вывод по графику: наиболее популярным среди разработчиков жанром является "Action"

In [ ]:
treemap_count_data = games_df[games_df['year'] >= YEAR_FROM].groupby(['platform', 'genre'], as_index=False)['global_sales'].count()

fig = px.treemap(
    data_frame=treemap_count_data,
    path=['platform', 'genre'],
    values='global_sales',
    height=800,
    width=1000,
    title=f'Распределение по количеству релизов игр по платформам и жанрам c {YEAR_FROM} года',
    labels=dict(labels="Жанр", global_sales="Количеество", parent='Платформа')
)
fig.show()

Вывод по графику: наиболее популярной среди разработчиков платформой является "DS"

In [ ]:
treemap_sum_data = games_df[games_df['year'] >= YEAR_FROM].groupby(['genre', 'platform'], as_index=False)['global_sales'].sum()
treemap_sum_data['global_sales'] = treemap_sum_data['global_sales'].apply(lambda x: round(x, 1))


fig = px.treemap(
    data_frame=treemap_sum_data,
    path=['genre', 'platform'],
    values='global_sales',
    height=800,
    width=1000,
    title=f'Распределение суммы продаж по жанрам и платформам c {YEAR_FROM} года',
    labels=dict(parent="Жанр", global_sales="Сумма МЛН($)", labels='Платформа')
    )
fig.show()

Вывод по графику: наиболее продаваемым жанром является "Action"

In [ ]:
treemap_sum_data = games_df[games_df['year'] >= YEAR_FROM].groupby(['platform', 'genre'], as_index=False)['global_sales'].sum()
treemap_sum_data['global_sales'] = treemap_sum_data['global_sales'].apply(lambda x: round(x, 1))


fig = px.treemap(
    data_frame=treemap_sum_data,
    path=['platform', 'genre'],
    values='global_sales',
    height=800,
    width=1000,
    title=f'Распределение суммы продаж по платформам и жанрам c {YEAR_FROM} года',
    labels=dict(labels="Жанр", global_sales="Сумма МЛН($)", parent='Платформа')
    )
fig.show()

Вывод по графику: видео игры, разработанные для платформы "PS3", наиболее продаваемые

# Итоги:

* Гипотезу о наличии жанра игр (возможно в связке с платформой), который продается лучше можно считать подтверждённой, так как весомую долю современного рынка видео игр (около 1/3) занимают игры жанра "Action", особенно успешно продающиеся для платформ PlayStation и Xbox. 

* В датасете все видео игр привязаны к единственному жанру, но стоит учесть тот факт, что существыет множество примеров успешных видео игр, которые представленны как минимум в 2 жанрах (например: Action-Adventure, Action-Sooter), и видео игры привязанные только к жанру Action вероятно могут являются представителями и других жанров, из чего следует, что при оценивании популярностии и продаваемости жанров, стоит брать во внимание жанры, входящие в ТОП-3 / ТОП-5 (Sports, Shooter, Misc, Role-Playing, Adventure).

* Так же стоит брать во внимание такое являние, как "пиратство", что может искажать показатели популярности некоторых платформ (например PC). 

* Цель: Hайти наиболее перспективный жанр игры для потенциальной разработки, можно считать выполненной, так как жанр "Action" является наиболее популярным и продаваемым на современном рынке видео игр.